src/Scene.cpp

```diff
@@ -51,30 +51,47 @@ bool Scene::HasAnyGeometrySelected() {
 void Scene::Render() const {
     glClearColor( 0.2f, 0.2f, 0.2f, 1.0f );
     glClear( GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT | GL_STENCIL_BUFFER_BIT );
+    glEnable( GL_DEPTH_TEST );
+    glEnable( GL_STENCIL_TEST );
 
     // Draw non stencil objects here
     for ( int i = 0; i < ( int ) geometries.size(); i++ ) {
         if ( !geometries[i]->IsSelectedGeometry() ) {
-
-            Shader* program = geometries[i]->program;
-            program->use();
-            glStencilMask( 0x00 );
-            program->setMat4( "Model", geometries[i]->ModelMatrix );
-            program->setMat4( "View", camera->ViewMatrix );
-            program->setMat4( "Projection", camera->ProjectionMatrix );
-
-
-            glBindVertexArray( geometries[i]->vertexArray );
-            glBindBuffer( GL_ELEMENT_ARRAY_BUFFER, geometries[i]->faceBuffer );
-            glDrawElements(
-                GL_TRIANGLES,      // mode
-                geometries[i]->faces.size(),    // count
-                GL_UNSIGNED_INT,   // type
-                ( void* )0         // element array buffer offset
-            );
-
+            if ( geometries[i]->renderMode == GL_LINE_STRIP ) {
+                Shader* program = geometries[i]->program;
+                geometries[i]->Load();
+                
+                program->use();
+                program->setMat4( "Model", geometries[i]->ModelMatrix );
+                program->setMat4( "View", glm::mat4(1.0f) );
+                program->setMat4( "Projection", glm::mat4(1.0f) );
+
+                glBindVertexArray( geometries[i]->vertexArray );
+                glBindBuffer( GL_ELEMENT_ARRAY_BUFFER, geometries[i]->vertexBuffer );
+                glDrawArrays(
+                    geometries[i]->renderMode, 
+                    0,
+                    geometries[i]->vertices.size()
+                );
+            } else {
```

Comment:
geometries[i]->Load() should be called once after the sketching is done, not every frame, so prob in the InputManager.

You should move both these render logic to seperate functions like Scene::RenderStencil() as Scene::Render() is getting too long and hard to read. (e.g., Scene::RenderSketch()) and if you make a separate Sketch geometry in the scene, then you can call that outside the for loop.

You're also going to want to arc-length parametrize (really small arc-length) the sketch so we can get rid of points really close to each other and duplicates. Plus we'll have the guarantee that each point is equidistant